File: clients/javascript/railvoya_client.js
/**
 * RailVoya JavaScript / Node.js API Client
 * Works in Node.js (v18+) and modern Browsers
 */

class RailVoyaClient {
  constructor(baseUrl = "http://127.0.0.1:8000/api/v1") {
    this.baseUrl = baseUrl.replace(/\/$/, "");
  }

  /**
   * Search available trains between two stations
   * @param {Object} options
   * @param {string} options.origin - 3-4 letter IR station code (e.g. 'NDLS')
   * @param {string} options.destination - 3-4 letter station code (e.g. 'BPL')
   * @param {string} options.date - Date in YYYY-MM-DD
   * @param {string} [options.quota='GN'] - Quota code (GN, TQ, LD, SS)
   */
  async searchTrains({ origin, destination, date, quota = "GN" }) {
    const params = new URLSearchParams({
      origin,
      destination,
      journey_date: date,
      quota,
    });
    const response = await fetch(`${this.baseUrl}/trains/search?${params.toString()}`);
    if (!response.ok) {
      throw new Error(`Search failed: ${response.status} ${response.statusText}`);
    }
    return await response.json();
  }

  /**
   * Check 10-Digit PNR Status & Berth Allocation
   * @param {string} pnrNumber - 10-digit Indian Railways PNR
   */
  async getPnrStatus(pnrNumber) {
    const response = await fetch(`${this.baseUrl}/pnr/${encodeURIComponent(pnrNumber)}`);
    if (!response.ok) {
      throw new Error(`PNR check failed: ${response.status} ${response.statusText}`);
    }
    return await response.json();
  }

  /**
   * Check Live Satellite Radar & RTIS Running Status
   * @param {string} trainNumber - 5-digit train number (e.g. '20901')
   * @param {string} [journeyDate] - Date in YYYY-MM-DD
   */
  async getRunningStatus(trainNumber, journeyDate) {
    const query = journeyDate ? `?journey_date=${encodeURIComponent(journeyDate)}` : "";
    const response = await fetch(`${this.baseUrl}/running-status/${encodeURIComponent(trainNumber)}${query}`);
    if (!response.ok) {
      throw new Error(`Running status failed: ${response.status} ${response.statusText}`);
    }
    return await response.json();
  }

  /**
   * Generate official IRCTC Fast-Track Handoff URL
   */
  getOfficialIrctcHandoff() {
    return "https://www.irctc.co.in/nget/train-search";
  }
}

// Export for Node.js CommonJS & ESM
if (typeof module !== "undefined" && module.exports) {
  module.exports = { RailVoyaClient };
}

// Example Execution when run directly via `node railvoya_client.js`
if (typeof require !== "undefined" && require.main === module) {
  (async () => {
    console.log("=== RailVoya JavaScript API Client Demo ===");
    const client = new RailVoyaClient();

    try {
      console.log("\n1. Searching Trains (NDLS ➔ BPL)...");
      const searchRes = await client.searchTrains({
        origin: "NDLS",
        destination: "BPL",
        date: new Date().toISOString().split("T")[0],
        quota: "GN",
      });
      const trains = Array.isArray(searchRes) ? searchRes : (searchRes.trains || []);
      console.log(`Found ${trains.length} trains.`);
      trains.slice(0, 2).forEach((t) => {
        console.log(`- ${t.train_number} ${t.train_name} (${t.departure_time} ➔ ${t.arrival_time})`);
      });

      console.log("\n2. Checking Live PNR (2458910243)...");
      const pnrRes = await client.getPnrStatus("2458910243");
      console.log(`PNR: ${pnrRes.pnr_number} | Train: ${pnrRes.train_number} | Status: ${pnrRes.chart_status}`);
      console.log(`Coach: ${pnrRes.passengers[0].coach}, Berth: ${pnrRes.passengers[0].berth_number}`);

      console.log("\n3. Official IRCTC Link:");
      console.log(client.getOfficialIrctcHandoff());
    } catch (err) {
      console.error("API Error:", err.message);
    }
  })();
}

File: clients/java/RailVoyaClient.java
package clients.java;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.LocalDate;
import java.time.Duration;

/**
 * RailVoya Java API Client
 * Enterprise Indian Railways integration client written in Modern Java (Java 11+)
 * Uses built-in java.net.http.HttpClient (Zero external dependencies).
 */
public class RailVoyaClient {

    private final String baseUrl;
    private final HttpClient httpClient;

    public RailVoyaClient() {
        this("http://127.0.0.1:8000/api/v1");
    }

    public RailVoyaClient(String baseUrl) {
        this.baseUrl = baseUrl.replaceAll("/$", "");
        this.httpClient = HttpClient.newBuilder()
                .version(HttpClient.Version.HTTP_2)
                .connectTimeout(Duration.ofSeconds(10))
                .build();
    }

    /**
     * Search available trains between two railway stations
     *
     * @param origin      Origin station code (e.g. "NDLS")
     * @param destination Destination station code (e.g. "BPL")
     * @param journeyDate Date in YYYY-MM-DD
     * @param quota       Quota code ("GN", "TQ", "LD", "SS")
     * @return JSON String of train search results
     */
    public String searchTrains(String origin, String destination, String journeyDate, String quota)
            throws IOException, InterruptedException {
        String endpoint = String.format("%s/trains/search?origin=%s&destination=%s&journey_date=%s&quota=%s",
                this.baseUrl,
                urlEncode(origin),
                urlEncode(destination),
                urlEncode(journeyDate),
                urlEncode(quota != null ? quota : "GN"));

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(endpoint))
                .timeout(Duration.ofSeconds(15))
                .header("Accept", "application/json")
                .header("User-Agent", "RailVoya-Java-SDK/1.0")
                .GET()
                .build();

        HttpResponse<String> response = this.httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() >= 400) {
            throw new RuntimeException("HTTP Error " + response.statusCode() + ": " + response.body());
        }
        return response.body();
    }

    /**
     * Check 10-digit Indian Railways PNR Status & Berth Allocation
     *
     * @param pnrNumber 10-digit PNR
     * @return JSON String with PNR status
     */
    public String getPnrStatus(String pnrNumber) throws IOException, InterruptedException {
        String endpoint = String.format("%s/pnr/%s", this.baseUrl, urlEncode(pnrNumber));

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(endpoint))
                .timeout(Duration.ofSeconds(15))
                .header("Accept", "application/json")
                .header("User-Agent", "RailVoya-Java-SDK/1.0")
                .GET()
                .build();

        HttpResponse<String> response = this.httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() >= 400) {
            throw new RuntimeException("HTTP Error " + response.statusCode() + ": " + response.body());
        }
        return response.body();
    }

    /**
     * Get Live Satellite Radar & RTIS GPS Running Status
     *
     * @param trainNumber 5-digit train number (e.g. "20901")
     * @return JSON String with running status
     */
    public String getRunningStatus(String trainNumber) throws IOException, InterruptedException {
        String endpoint = String.format("%s/running-status/%s", this.baseUrl, urlEncode(trainNumber));

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(endpoint))
                .timeout(Duration.ofSeconds(15))
                .header("Accept", "application/json")
                .header("User-Agent", "RailVoya-Java-SDK/1.0")
                .GET()
                .build();

        HttpResponse<String> response = this.httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() >= 400) {
            throw new RuntimeException("HTTP Error " + response.statusCode() + ": " + response.body());
        }
        return response.body();
    }

    /**
     * Get Official IRCTC Fast-Track Deep Link
     */
    public String getOfficialIrctcUrl() {
        return "https://www.irctc.co.in/nget/train-search";
    }

    private static String urlEncode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    /**
     * Standalone main runner demo
     */
    public static void main(String[] args) {
        System.out.println("==========================================");
        System.out.println("  RailVoya Java SDK (Java 11+ HttpClient) ");
        System.out.println("==========================================");

        RailVoyaClient client = new RailVoyaClient();
        String tomorrow = LocalDate.now().plusDays(1).toString();

        try {
            System.out.println("\n[1] Calling Train Search API (NDLS ➔ BPL for " + tomorrow + ")...");
            String searchJson = client.searchTrains("NDLS", "BPL", tomorrow, "GN");
            System.out.println("Response Payload:\n" + truncate(searchJson, 250));

            System.out.println("\n[2] Calling CRIS PNR Status API (PNR #2458910243)...");
            String pnrJson = client.getPnrStatus("2458910243");
            System.out.println("Response Payload:\n" + truncate(pnrJson, 250));

            System.out.println("\n[3] Calling Live RTIS Running Status (Train 20901 Vande Bharat)...");
            String runningJson = client.getRunningStatus("20901");
            System.out.println("Response Payload:\n" + truncate(runningJson, 250));

            System.out.println("\n[4] Official IRCTC Handoff Gateway:");
            System.out.println(client.getOfficialIrctcUrl());

            System.out.println("\n>> Java SDK Execution Completed Successfully!");

        } catch (Exception e) {
            System.err.println("Execution Error: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static String truncate(String text, int maxLen) {
        if (text == null || text.length() <= maxLen) return text;
        return text.substring(0, maxLen) + "... [truncated]";
    }
}
